tests/ui-api-layer-acceptance-tests/graphql/idtokenprovider.go

```
@@ -0,0 +1,114 @@
+	if strings.Contains(loginEndpoint, "#.*error") {
+		return nil, errors.New(fmt.Sprintf("Login - Redirected with error: '%s'", loginEndpoint))
+	}
+
+	_, err2 := p.httpClient.Get(p.config.dexConfig.baseUrl + loginEndpoint)
+	if err2 != nil {
+		return nil, err2
+	}
+
+	loginResp, err3 := p.httpClient.PostForm(p.config.dexConfig.baseUrl+loginEndpoint, url.Values{
+		"login":    {p.config.userCredentials.username},
+		"password": {p.config.userCredentials.password},
+	})
+	if err3 != nil {
+		return nil, err3
+	}
+	if loginResp.StatusCode < 300 || loginResp.StatusCode > 399 {
+		return nil, errors.New(fmt.Sprintf("Login - response error: '%s' - %s", loginResp.Status, readRespBody(loginResp)))
+	}
+
+	// /approval?req=qruhpy2cqjvv4hcrbuu44mf4v
+	approvalEndpoint := loginResp.Header.Get("location")
+	if strings.Contains(approvalEndpoint, "#.*error") {
+		return nil, errors.New(fmt.Sprintf("Approval - Redirected with error: '%s'", approvalEndpoint))
+	}
+	approvalResp, err4 := p.httpClient.Get(p.config.dexConfig.baseUrl + approvalEndpoint)
+	if err4 != nil {
+		return nil, err4
+	}
+	if approvalResp.StatusCode < 300 || approvalResp.StatusCode > 399 {
+		return nil, errors.New(fmt.Sprintf("Approval - response error: '%s' - %s", approvalResp.Status, readRespBody(approvalResp)))
+	}
+
+	clientEndpoint := approvalResp.Header.Get("location")
+	if strings.Contains(clientEndpoint, "#.*error") {
+		return nil, errors.New(fmt.Sprintf("Client - Redirected with error: '%s'", clientEndpoint))
+	}
+
+	parsedUrl, parseErr := url.Parse(clientEndpoint)
+	if parseErr != nil {
+		return nil, parseErr
+	}
+
+	var result map[string]string = make(map[string]string)
+	fragmentParams := strings.Split(parsedUrl.Fragment, "&")
+	for _, param := range fragmentParams {
+		keyAndValue := strings.Split(param, "=")
+		result[keyAndValue[0]] = keyAndValue[1]
+	}
+
+	return result, nil
+}
+
+func readRespBody(resp *http.Response) string {
+	defer resp.Body.Close()
+	b, _ := ioutil.ReadAll(resp.Body)
```

Please do not ignore error

---

done